src/udev/dmi_memory_id/dmi_memory_id.c

```diff
@@ -0,0 +1,808 @@
+ *    http://www.dmtf.org/standards/cim/
+ *  - IPMI 2.0 revision 1.0
+ *    "Intelligent Platform Management Interface Specification"
+ *    http://developer.intel.com/design/servers/ipmi/spec.htm
+ *  - AMD publication #25481 revision 2.28
+ *    "CPUID Specification"
+ *    http://www.amd.com/us-en/assets/content_type/white_papers_and_tech_docs/25481.pdf
+ *  - BIOS Integrity Services Application Programming Interface version 1.0
+ *    http://www.intel.com/design/archives/wfm/downloads/bisspec.htm
+ *  - DMTF DSP0239 version 1.1.0
+ *    "Management Component Transport Protocol (MCTP) IDs and Codes"
+ *    http://www.dmtf.org/standards/pmci
+ *  - "TPM Main, Part 2 TPM Structures"
+ *    Specification version 1.2, level 2, revision 116
+ *    https://trustedcomputinggroup.org/tpm-main-specification/
+ *  - "PC Client Platform TPM Profile (PTP) Specification"
+ *    Family "2.0", Level 00, Revision 00.43, January 26, 2015
+ *    https://trustedcomputinggroup.org/pc-client-platform-tpm-profile-ptp-specification/
+ *  - "RedFish Host Interface Specification" (DMTF DSP0270)
+ *    https://www.dmtf.org/sites/default/files/DSP0270_1.0.1.pdf
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <inttypes.h>
+#include <errno.h>
+
+#include "alloc-util.h"
+#include "build.h"
+#include "fileio.h"
+#include "main-func.h"
+#include "unaligned.h"
+#include "udev-util.h"
+
+#define SUPPORTED_SMBIOS_VER 0x030300
+
+static const char *arg_source_file = NULL;
+
+#define OUT_OF_SPEC_STR "<OUT OF SPEC>"
+#define SYS_FIRMWARE_DIR "/sys/firmware/dmi/tables"
+#define SYS_ENTRY_FILE SYS_FIRMWARE_DIR "/smbios_entry_point"
+#define SYS_TABLE_FILE SYS_FIRMWARE_DIR "/DMI"
+
+/* Use memory alignment workaround or not */
+#ifdef __ia64__
+#define ALIGNMENT_WORKAROUND
+#endif
+
+struct dmi_header
+{
+        uint8_t type;
+        uint8_t length;
+        uint16_t handle;
+        uint8_t *data;
+};
+
+/*
+ * You may use the following defines to adjust the type definitions
+ * depending on the architecture:
+ * - Define ALIGNMENT_WORKAROUND if your system doesn't support
+ *   non-aligned memory access. In this case, we use a slower, but safer,
+ *   memory access method. This should be done automatically in config.h
+ *   for architectures which need it.
+ */
+
+#if __BYTE_ORDER == __BIG_ENDIAN
+typedef struct {
+        uint32_t h;
+        uint32_t l;
+} ULargeInteger;
+#else
+typedef struct {
+        uint32_t l;
+        uint32_t h;
+} ULargeInteger;
+#endif
+
+static inline ULargeInteger to_ularge_integer(uint32_t low, uint32_t high)
+{
+        ULargeInteger self;
+
+        self.l = low;
+        self.h = high;
+
+        return self;
+}
+
+/*
+ * Per SMBIOS v2.8.0 and later, all structures assume a little-endian
+ * ordering convention.
+ */
+#define WORD(x) (unaligned_read_le16(x))
+#define DWORD(x) (unaligned_read_le32(x))
+#define QWORD(x) (to_ularge_integer(DWORD(x), DWORD(x + 4)))
```

Comment:
why retain uLargeInteger? Isn't that just a remnant from really old C compilers which didn't have 64bit integers?

this should just use unaligned_read_le64() here, no? anything special about this that requires other accessors?

Reply:
(if ULargeInteger has some value (which i don't see) then it should probably be marked with `_packet_`)